HW2/P2/parallel_vector.pyx

```diff
@@ -54,62 +54,83 @@ cpdef move_data_serial(np.int32_t[:] counts,
                        np.int32_t[:] src,
                        np.int32_t[:] dest,
                        int repeat):
-   cdef:
-       int idx, r
+    cdef:
+        int idx, r
 
-   assert src.size == dest.size, "Sizes of src and dest arrays must match"
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+    assert src.size == dest.size, "Sizes of src and dest arrays must match"
+    with nogil:
+        for r in range(repeat):
+            for idx in range(src.shape[0]):
+                # Assumption: Code here does not need to be changed
+                # for coarse-grained locking, one lock should be used
+                if counts[src[idx]] > 0:
+                    counts[dest[idx]] += 1
+                    counts[src[idx]] -= 1
 
 
 cpdef move_data_fine_grained(np.int32_t[:] counts,
                              np.int32_t[:] src,
                              np.int32_t[:] dest,
                              int repeat):
-   cdef:
-       int idx, r
-       omp_lock_t *locks = get_N_locks(counts.shape[0])
-
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for each element of counts to parallelize
-   # data movement.  Be sure to avoid deadlock, and double-locking.
-   ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
-
-   free_N_locks(counts.shape[0], locks)
-
+    cdef:
+        int idx, r
+        omp_lock_t * locks = get_N_locks(counts.shape[0])
+        int thread_count = 4  # omp_get_num_threads()
+
+    ##########
+    # Your code here
+    # Use parallel.prange() and a lock for each element of counts to parallelize
+    # data movement.  Be sure to avoid deadlock, and double-locking.
+    ##########
```

Comment:
@samighoche is correct.  Another way to do it is to check the condition again after acquiring the lock (since the condition may have changed after acquiring the lock.